cmd/juju/commands/upgrademodel.go

```diff
@@ -396,14 +403,328 @@ func (c *upgradeJujuCommand) Run(ctx *cmd.Context) (err error) {
+
+	// If target version is the current local binary version, then try to upload.
+	canImplicitUpload := CheckCanImplicitUpload(
+		modelType, isOfficialClient(), jujuversion.Current, agentVersion,
+	)
+	if !canImplicitUpload || !isControllerModel {
+		// expecting to upload a local binary but we are not allowed to upload, so pretend there
+		// is no more recent version available.
+		logger.Debugf("no available binary found, and we are not allowed to upload, err %v", err)
+		return chosenVersion, errUpToDate
+	}
+
+	if targetVersion.Compare(jujuversion.Current.ToPatch()) != 0 {
+		logger.Warningf(
+			"try again with --agent-version=%s if you want to upload the local binary",
+			jujuversion.Current.ToPatch(),
+		)
+		return chosenVersion, errUpToDate
+	}
+
+	// found a best target version but a local binary is required to be uploaded.
+	if chosenVersion, err = uploadTools(modelUpgrader, false, agentVersion, dryRun); err != nil {
+		return chosenVersion, block.ProcessBlockedError(err, block.BlockChange)
+	}
+	fmt.Fprintf(ctx.Stdout,
+		"no prepackaged agent binaries available, using local agent binary %v%s\n",
+		chosenVersion, "",
+	)
+
+	chosenVersion, err = c.notifyControllerUpgrade(ctx, modelUpgrader, chosenVersion, dryRun)
+	if err != nil {
+		return chosenVersion, err
+	}
+	return chosenVersion, nil
+}
+
+func (c *upgradeJujuCommand) upgradeModel(
+	ctx *cmd.Context, modelUpgrader ModelUpgraderAPI, fetchTimeout time.Duration,
+	modelType model.ModelType,
+) (err error) {
+	targetVersion := c.Version
+	defer func() {
+		if err == nil {
+			fmt.Fprintf(ctx.Stderr, "best version:\n    %v\n", targetVersion)
+			if c.DryRun {
+				if c.BuildAgent {
+					fmt.Fprintf(ctx.Stderr, "%s --build-agent\n", c.upgradeMessage)
+				} else {
+					fmt.Fprintf(ctx.Stderr, "%s\n", c.upgradeMessage)
+				}
+			} else {
+				fmt.Fprintf(ctx.Stdout, "started upgrade to %s\n", targetVersion)
+			}
+		}
+
+		if err == errUpToDate {
+			ctx.Infof(err.Error())
+			err = nil
+		}
+		if err != nil {
+			logger.Debugf("upgradeModel failed %v", err)
+		}
+	}()
+
+	controllerClient, err := c.getControllerAPI()
+	if err != nil {
+		return err
+	}
+	defer controllerClient.Close()
+
+	modelConfigClient, err := c.getModelConfigAPI()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer modelConfigClient.Close()
+
+	attrs, err := modelConfigClient.ModelGet()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	cfg, err := config.New(config.NoDefaults, attrs)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	agentVersion, ok := cfg.AgentVersion()
+	if !ok {
+		// Can't happen. In theory.
+		return errors.New("incomplete model configuration")
+	}
+	if majorDiff := jujuversion.Current.Major - agentVersion.Major; majorDiff != 0 && majorDiff != 1 {
+		// This version of juju client cannot upgrade the running
+		// model version (can't guarantee API compatibility).
+		return errors.Errorf("cannot upgrade a %s model with a %s client", agentVersion, jujuversion.Current)
+	}
```

Comment:
We don't need this check anymore. We check the facade versions etc - that's all we need.